src/libertem/executor/pipelined.py

```diff
@@ -0,0 +1,857 @@
+
+        # timeout for cleanup, either from exception or when joining processes
+        self._cleanup_timeout = cleanup_timeout
+
+        # timeout for starting a single worker
+        self._startup_timeout = startup_timeout
+
+        # keep this at the bottom:
+        self._pool = self._start_pool()
+
+    def _start_pool(self) -> WorkerPool:
+        with tracer.start_as_current_span("PipelinedExecutor.start_pool") as span:
+            pool = WorkerPool(
+                worker_fn=functools.partial(
+                    pipelined_worker,
+                    pin=self._pin_workers,
+                    early_setup=self._early_setup,
+                ),
+                spec=self._spec,
+            )
+            # FIXME: check process liveness here, too, to catch early startup
+            # failures without having to run into the timeout
+            for i in range(pool.size):
+                with pool.response_queue.get(timeout=self._startup_timeout) as (msg, _):
+                    if msg["type"] == "ERROR":
+                        raise RuntimeError(
+                            f"error on startup: {msg['error']}"
+                        )
+                    if msg["type"] != "STARTUP_DONE":
+                        raise RuntimeError(
+                            f"unknown message type {msg['type']}, expected STARTUP_DONE"
+                        )
+                    span.add_event("worker startup done", {"worker_id": msg["worker_id"]})
+            # set here, so we don't try to close the pool if it doesn't exist
+            self._closed = False
+            return pool
+
+    @classmethod
+    def _default_spec(cls):
+        from libertem.utils.devices import detect
+        detected = detect()
+        return _make_spec(**detected)
+
+    @classmethod
+    def make_local(cls, **kwargs):
+        """
+        Create a :code:`PipelinedExecutor` with the default spec.
+        """
+        spec = cls._default_spec()
+        return cls(spec=spec, **kwargs)
+
+    @classmethod
+    def make_spec(cls, *args, **kwargs):
+        return _make_spec(*args, **kwargs)
+    make_spec.__doc__ = _make_spec.__doc__
+
+    def _validate_worker_state(self):
+        if not self._pool.all_alive():
+            raise RuntimeError("some workers are stopped, cannot continue")
+
+    def _run_tasks_inner(
+        self,
+        tasks: Iterable[TaskProtocol],
+        params_handle: Any,
+        cancel_id: Any,
+        task_comm_handler: "TaskCommHandler",
+    ) -> ResultWithID:
+        in_flight = 0
+        id_to_task = {}
+
+        try:
+            self._validate_worker_state()
+            task_comm_handler.start()
+            span = trace.get_current_span()
+            span_context = span.get_span_context()
+
+            for qs, _ in self._pool.workers:
+                qs.request.put({
+                    "type": "WARMUP",
+                    "span_context": span_context,
+                })
```

Comment:
Good point - I don't think it has to run for every `run_udf` - I'll move this to `_start_pool` to also properly attribute the timings to the startup phase. Yeah, `import`ing things that are already in `sys.modules` is very cheap (<100ns), the `with env.enter()` can be a bit more expensive though.

I think I put it in `run_tasks` because I also wanted to warm up the whole I/O pipeline, in case we have `numba` compiled code in there. This of course only works if we know which `DataSet` / `Acquisition` instance will be used. I didn't get to implement this yet, so i'll defer this for now and move warmup to `_start_pool`.

Reply:
WARMUP is now moved to `_start_pool`